### DISCUSSION TITLE
Custom logic for (de)serialization of eloquent models

### DISCUSSION THREAD

**Discussion by dennisprudlo** [livewire/livewire#4620] [3. Feature ideas]
Hey guys,

When using full page components it's very handy to use the route model binding implementation of livewire. My models have (additional to a default id) stripe like unique IDs which I use for routes and other client stuff, so the internal ID won't be exposed.

```html
example.com/users/user_LkzQ3Y2JUL0YyiSvwzlzhibT

<select>
    <option value="user_LkzQ3Y2JUL0YyiSvwzlzhibT">John Doe</option>
   ...
</select>
```

However when using full page components (or just any other situation you leave livewire with handling the (de)serialization of a model) you end up with the model name and the internal ID in the livewire payload.

In "upsert" cases (where I use the same component for creating and editing a resource) I created my own `OptionalModel` wireable class which I use where I only store the UID (`user_LkzQ3Y2JUL0YyiSvwzlzhibT`) and when receiving this from a livewire request I know the resource (`user`) and the actual record (`LkzQ3Y2JUL0YyiSvwzlzhibT`).

It would be great if we could define our own logic for the model serialization or at least what values to use for the `class` and `id` fields in the model object of a livewire request
```json
"models": {
    "user": {
        "class": "App\\Models\\User",  <--
        "id": 167,                     <--
        "relations": [ "..." ],
        "connection": "..."
    }
}
```

## Workaround
As a (dirty) workaround I receive the model in the mount-method and convert it to said `OptionalModel` wireable and use this one further on. But theres two annoying things here:
1. I can't name my OptionalModel property `$user` because livewire will then try to put the model from the mount method into that property, but that won't work so I have to rename the property to something different like `$optionalUser`. I could rename the route model binding var to something different like `userModel` but that seems odd for the same reason.
2. I have to always pass the actual model to the view with the expected name so the view can access the user with `$user` (Otherwise I would have to use `$optionalUser->model` everytime inside the view).

```php
// Route example
Route::get('{user}', User::class);

// Class Example
class User extends Component {

    public OptionalModel $optionalUser;

    public function mount (\App\Models\User $user) : void {
        $this->optionalUser = new OptionalModel($user::class, $user->secret);
    }

    public function render () : View {
        return view('pages.user', [
            'user' => $this->optionalUser->model
        ]);
    }
}
```

What do you think?

**Comment by joshhanley** [comment]
@dennisprudlo yeah there is a PR in progress #4154 which will allow you to write your own property handlers. Will that allow you to do what you are hoping?

**Reply by dennisprudlo** [reply]
Oh yeah that seems like something I could work with :) another example with property attributes (https://github.com/livewire/livewire/pull/4565) would be perfect as I wouldn't have to reimplement the logic for the connection, relations and so on. Will keep an eye on those PRs

**Reply by joshhanley** [reply]
@dennisprudlo ok good to know 🙂
